src/utils.c

```diff
@@ -444,12 +444,55 @@ void print_image(FILE *fptr)
         printf("%s",read_string);
 }
 
-double koinu_to_coins(uint64_t koinu) {
-    return (double)koinu / (double)100000000;
+long double koinu_to_coins(uint64_t koinu) {
+    return (long double)koinu / (long double)1e8;
 }
 
-uint64_t coins_to_koinu(float coins) {
-    return ((float)coins * (float)100000000);
+uint64_t coins_to_koinu(long double coins) {
+    char coins_string[32];
+    char koinu_string[32];
+    sprintf(coins_string, "%.9Lf", coins);
+    char* c_ptr = coins_string;
+    char* k_ptr = koinu_string;
+
+    //copy all digits until end of string or decimal point is reached
+    while (*c_ptr != '\0') {
+        if (*c_ptr =='.') {
+            c_ptr++;
+            break;
+        }
+        memcpy(k_ptr, c_ptr, 1);
+        c_ptr++;
+        k_ptr++;
+    }
+    //copy remaining 8 or less decimal places
+    for (int i=1; i<=8; i++) {
+        if (*c_ptr=='\0') {
+            memset(k_ptr, 0, 8-i); //fill with zeroes if less than 8 decimal places
+            k_ptr+=(8-i);
+            memset(k_ptr, '\0', 1);
+            break;
+        }
+        if (i==8) {
+            //assessing the 9th decimal place
+            if (c_ptr[1] >= '5') {
```

Comment:
leaving note to explore why this fails on amd64 when running in valgrind with:
`make && valgrind  --leak-check=full --show-leak-kinds=all --track-origins=yes --error-exitcode=1 ./tests`